File: zig/src/format.zig
//! The on-disk container format (magic "DRDO", version 4; version 3 still read):
//! header marshalling and parse. Matches the Rust reference and the other ports
//! byte-for-byte. All multi-byte integers are big-endian.

const std = @import("std");
const mem = std.mem;
const tf = @import("threefish.zig");

pub const Variant = tf.Variant;

pub const MAGIC = "DRDO";
pub const VERSION = 4;
pub const DEFAULT_CHUNK_BYTES: u32 = 64 * 1024;
pub const MAX_CHUNK_BYTES: u32 = 1 << 30;
pub const MAX_LABEL_LEN = 4096;
pub const MAC_KEY_LEN = 32;
pub const TAG_LEN = 32;

pub const Mac = enum(u8) {
    hmac = 1,
    skein = 2,
    blake3 = 3,

    pub fn fromCode(c: u8) ?Mac {
        return switch (c) {
            1 => .hmac,
            2 => .skein,
            3 => .blake3,
            else => null,
        };
    }
};

pub const KdfKind = enum(u8) {
    argon2id = 1,
    scrypt = 2,
    pbkdf2 = 3,
};

pub const Kdf = union(KdfKind) {
    argon2id: struct { m_cost: u32, t_cost: u32, p_cost: u32 },
    scrypt: struct { log_n: u8, r: u32, p: u32 },
    pbkdf2: struct { rounds: u32 },

    pub fn mkArgon2id(m_cost_kib: u32, t_cost: u32, p_cost: u32) Kdf {
        return .{ .argon2id = .{ .m_cost = m_cost_kib, .t_cost = t_cost, .p_cost = p_cost } };
    }
    pub fn mkScrypt(log_n: u8, r: u32, p: u32) Kdf {
        return .{ .scrypt = .{ .log_n = log_n, .r = r, .p = p } };
    }
    pub fn mkPbkdf2(rounds: u32) Kdf {
        return .{ .pbkdf2 = .{ .rounds = rounds } };
    }
};

pub const Error = error{
    BadMagic,
    UnsupportedVersion,
    UnknownVariant,
    UnknownKdf,
    UnknownPrf,
    UnknownMac,
    LabelTooLong,
    Truncated,
};

pub const Header = struct {
    version: u8,
    variant: Variant,
    kdf: Kdf,
    mac: Mac,
    chunk_size: u32,
    salt: [256]u8,
    salt_len: usize,
    tweak: [16]u8,
    iv: [128]u8,
    label: [MAX_LABEL_LEN]u8,
    label_len: usize,

    pub fn ivSlice(self: *const Header) []const u8 {
        return self.iv[0..self.variant.keyLen()];
    }
    pub fn labelSlice(self: *const Header) []const u8 {
        return self.label[0..self.label_len];
    }
    pub fn saltSlice(self: *const Header) []const u8 {
        return self.salt[0..self.salt_len];
    }
};

/// Serialize a header into out, returning the number of bytes written.
pub fn marshal(h: *const Header, out: []u8) usize {
    var n: usize = 0;
    @memcpy(out[0..4], MAGIC);
    n += 4;
    out[n] = h.version;
    n += 1;
    out[n] = @intFromEnum(h.variant);
    n += 1;
    switch (h.kdf) {
        .argon2id => |a| {
            out[n] = 1;
            n += 1;
            mem.writeInt(u32, out[n..][0..4], a.m_cost, .big);
            n += 4;
            mem.writeInt(u32, out[n..][0..4], a.t_cost, .big);
            n += 4;
            mem.writeInt(u32, out[n..][0..4], a.p_cost, .big);
            n += 4;
        },
        .scrypt => |s| {
            out[n] = 2;
            n += 1;
            out[n] = s.log_n;
            n += 1;
            mem.writeInt(u32, out[n..][0..4], s.r, .big);
            n += 4;
            mem.writeInt(u32, out[n..][0..4], s.p, .big);
            n += 4;
        },
        .pbkdf2 => |p| {
            out[n] = 3;
            n += 1;
            mem.writeInt(u32, out[n..][0..4], p.rounds, .big);
            n += 4;
            out[n] = 1; // prf id = HMAC-SHA256
            n += 1;
        },
    }
    out[n] = @intFromEnum(h.mac);
    n += 1;
    mem.writeInt(u32, out[n..][0..4], h.chunk_size, .big);
    n += 4;
    out[n] = @intCast(h.salt_len);
    n += 1;
    @memcpy(out[n .. n + h.salt_len], h.saltSlice());
    n += h.salt_len;
    @memcpy(out[n .. n + 16], &h.tweak);
    n += 16;
    const iv_len = h.variant.keyLen();
    @memcpy(out[n .. n + iv_len], h.ivSlice());
    n += iv_len;
    if (h.version >= 4) {
        mem.writeInt(u16, out[n..][0..2], @intCast(h.label_len), .big);
        n += 2;
        @memcpy(out[n .. n + h.label_len], h.labelSlice());
        n += h.label_len;
    }
    return n;
}

/// A minimal byte-reader abstraction so the parser works over a file or a buffer.
pub const Reader = struct {
    ctx: *anyopaque,
    readFn: *const fn (ctx: *anyopaque, buf: []u8) usize,

    pub fn readFull(self: Reader, buf: []u8) bool {
        var got: usize = 0;
        while (got < buf.len) {
            const n = self.readFn(self.ctx, buf[got..]);
            if (n == 0) return false;
            got += n;
        }
        return true;
    }
};

fn rdU8(r: Reader) Error!u8 {
    var b: [1]u8 = undefined;
    if (!r.readFull(&b)) return Error.Truncated;
    return b[0];
}

fn rdU16(r: Reader) Error!u16 {
    var b: [2]u8 = undefined;
    if (!r.readFull(&b)) return Error.Truncated;
    return mem.readInt(u16, &b, .big);
}

fn rdU32(r: Reader) Error!u32 {
    var b: [4]u8 = undefined;
    if (!r.readFull(&b)) return Error.Truncated;
    return mem.readInt(u32, &b, .big);
}

/// Read a header from r, leaving it positioned at the first frame.
pub fn read(r: Reader, h: *Header) Error!void {
    var magic: [4]u8 = undefined;
    if (!r.readFull(&magic)) return Error.Truncated;
    if (!mem.eql(u8, &magic, MAGIC)) return Error.BadMagic;
    h.version = try rdU8(r);
    if (h.version != 3 and h.version != VERSION) return Error.UnsupportedVersion;
    h.variant = tf.Variant.fromCode(try rdU8(r)) orelse return Error.UnknownVariant;

    const kdf_id = try rdU8(r);
    switch (kdf_id) {
        1 => h.kdf = Kdf.mkArgon2id(try rdU32(r), try rdU32(r), try rdU32(r)),
        2 => h.kdf = Kdf.mkScrypt(try rdU8(r), try rdU32(r), try rdU32(r)),
        3 => {
            const rounds = try rdU32(r);
            if (try rdU8(r) != 1) return Error.UnknownPrf;
            h.kdf = Kdf.mkPbkdf2(rounds);
        },
        else => return Error.UnknownKdf,
    }

    h.mac = Mac.fromCode(try rdU8(r)) orelse return Error.UnknownMac;
    h.chunk_size = try rdU32(r);
    h.salt_len = try rdU8(r);
    if (!r.readFull(h.salt[0..h.salt_len])) return Error.Truncated;
    if (!r.readFull(&h.tweak)) return Error.Truncated;
    if (!r.readFull(h.iv[0..h.variant.keyLen()])) return Error.Truncated;
    if (h.version >= 4) {
        const label_len = try rdU16(r);
        if (label_len > MAX_LABEL_LEN) return Error.LabelTooLong;
        if (!r.readFull(h.label[0..label_len])) return Error.Truncated;
        h.label_len = label_len;
    } else {
        h.label_len = 0;
    }
}

File: cpp/src/format.cpp
#include "dorado/format.hpp"

#include <stdexcept>

namespace dorado::format {
namespace {

void put_u16(std::vector<std::uint8_t>& o, std::uint16_t w) {
  o.push_back(std::uint8_t(w >> 8));
  o.push_back(std::uint8_t(w));
}

void put_u32(std::vector<std::uint8_t>& o, std::uint32_t w) {
  for (int i = 3; i >= 0; --i) o.push_back(std::uint8_t(w >> (8 * i)));
}

// A cursor that throws on underflow; parse() catches and returns std::unexpected.
struct Cursor {
  std::span<const std::uint8_t> data;
  std::size_t pos = 0;
  std::span<const std::uint8_t> take(std::size_t n) {
    if (pos + n > data.size()) throw std::runtime_error("unexpected end of input");
    auto s = data.subspan(pos, n);
    pos += n;
    return s;
  }
  std::uint8_t u8() { return take(1)[0]; }
  std::uint16_t u16() {
    auto s = take(2);
    return std::uint16_t(std::uint16_t(s[0]) << 8 | s[1]);
  }
  std::uint32_t u32() {
    auto s = take(4);
    std::uint32_t v = 0;
    for (int i = 0; i < 4; ++i) v = v << 8 | s[i];
    return v;
  }
};

kdf::Kdf parse_kdf(std::uint8_t id, Cursor& c) {
  switch (id) {
    case 1: {
      std::uint32_t m = c.u32(), t = c.u32(), p = c.u32();
      return kdf::Argon2id{m, t, p};
    }
    case 2: {
      std::uint8_t logn = c.u8();
      std::uint32_t r = c.u32(), p = c.u32();
      return kdf::Scrypt{logn, r, p};
    }
    case 3: {
      std::uint32_t rounds = c.u32();
      std::uint8_t prf = c.u8();
      if (prf != 1) throw std::runtime_error("unknown pbkdf2 prf id");
      return kdf::Pbkdf2{rounds};
    }
    default:
      throw std::runtime_error("unknown kdf id");
  }
}

}  // namespace

std::uint8_t variant_code(Variant v) {
  switch (v) {
    case Variant::TF256: return 0;
    case Variant::TF512: return 1;
    case Variant::TF1024: return 2;
  }
  return 0;
}

std::vector<std::uint8_t> serialize(const Header& h) {
  std::vector<std::uint8_t> o = {'D', 'R', 'D', 'O'};
  o.push_back(h.version);
  o.push_back(variant_code(h.variant));
  std::visit(
      [&](const auto& k) {
        using T = std::decay_t<decltype(k)>;
        if constexpr (std::is_same_v<T, kdf::Argon2id>) {
          o.push_back(1);
          put_u32(o, k.m_cost);
          put_u32(o, k.t_cost);
          put_u32(o, k.p_cost);
        } else if constexpr (std::is_same_v<T, kdf::Scrypt>) {
          o.push_back(2);
          o.push_back(k.log_n);
          put_u32(o, k.r);
          put_u32(o, k.p);
        } else {
          o.push_back(3);
          put_u32(o, k.rounds);
          o.push_back(1);  // prf id = HMAC-SHA256
        }
      },
      h.kdf);
  o.push_back(mac::mac_id(h.mac));
  put_u32(o, h.chunk_size);
  o.push_back(std::uint8_t(h.salt.size()));
  o.insert(o.end(), h.salt.begin(), h.salt.end());
  o.insert(o.end(), h.tweak.begin(), h.tweak.end());
  o.insert(o.end(), h.iv.begin(), h.iv.end());
  if (h.version >= 4) {
    put_u16(o, std::uint16_t(h.label.size()));
    o.insert(o.end(), h.label.begin(), h.label.end());
  }
  return o;
}

std::expected<ParseResult, std::string> parse(std::span<const std::uint8_t> input) {
  try {
    Cursor c{input};
    auto magic = c.take(4);
    if (!(magic[0] == 'D' && magic[1] == 'R' && magic[2] == 'D' && magic[3] == 'O'))
      return std::unexpected("not a dorado container (bad magic)");
    Header h;
    h.version = c.u8();
    if (h.version != 3 && h.version != 4) return std::unexpected("unsupported version");
    std::uint8_t vcode = c.u8();
    switch (vcode) {
      case 0: h.variant = Variant::TF256; break;
      case 1: h.variant = Variant::TF512; break;
      case 2: h.variant = Variant::TF1024; break;
      default: return std::unexpected("unknown variant code");
    }
    h.kdf = parse_kdf(c.u8(), c);
    auto m = mac::mac_from_id(c.u8());
    if (!m) return std::unexpected("unknown mac id");
    h.mac = *m;
    h.chunk_size = c.u32();
    std::uint8_t salt_len = c.u8();
    auto salt = c.take(salt_len);
    h.salt.assign(salt.begin(), salt.end());
    auto tweak = c.take(16);
    h.tweak.assign(tweak.begin(), tweak.end());
    auto iv = c.take(block_size(h.variant));
    h.iv.assign(iv.begin(), iv.end());
    if (h.version >= 4) {
      std::uint16_t ll = c.u16();
      auto label = c.take(ll);
      h.label.assign(label.begin(), label.end());
    }
    return ParseResult{std::move(h), c.pos};
  } catch (const std::exception& e) {
    return std::unexpected(e.what());
  }
}

}  // namespace dorado::format
